Classes/Domain/NodeCreation/NodeCreationElements.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Neos\Neos\Ui\Domain\NodeCreation;
+
+use Neos\ContentRepository\Core\SharedModel\Node\NodeAggregateIds;
+
+/**
+ * Access to deserialize elements of the node creation dialog
+ *
+ * Property-like elements are of simple types or objects.
+ * The values will be deserialized according to its type.
+ * For example myImage will be an actual image object instance.
+ *
+ *     Vendor.Site:Content:
+ *       ui:
+ *         creationDialog:
+ *           elements:
+ *             myString:
+ *               type: string
+ *             myImage:
+ *               type: Neos\Media\Domain\Model\ImageInterface
+ *
+ * Reference-like elements are of type `references` or `reference`
+ * And will be available as NodeAggregateIds collection.
+ *
+ *     Vendor.Site:Content:
+ *       ui:
+ *         creationDialog:
+ *           elements:
+ *             myReferences:
+ *               type: references
+ *
+ * The same categories apply to promoted elements:
+ *
+ *     Vendor.Site:Content:
+ *       properties:
+ *         myString:
+ *           type: string
+ *           ui:
+ *             showInCreationDialog: true
+ *
+ * @api As part of the {@see NodeCreationHandlerInterface}
+ */
+final readonly class NodeCreationElements
+{
+    /**
+     * @param array<string, mixed> $propertyLikeValues
+     * @param array<string, NodeAggregateIds> $referenceLikeValues
+     * @param array<int|string, mixed> $serializedValues
+     * @internal you should not need to construct this
+     */
+    public function __construct(
+        private array $propertyLikeValues,
+        private array $referenceLikeValues,
```

Comment:
```suggestion
        private array $elementValues,
```

array<string, NodeAggregateIds|mixed>

Reply:
as far as I understand it, this DTO is an object that can be used to prime the creation of nodes with some parameters and we (currently) use it to map the values of the NodeCreation dialog.

As mentioned in our weekly today, I wonder why we need to separate the values already at this point. If we _do_ I would suggest `$propertyValues` and `$referencedNodeAggregateIds` but I would personally prefer to keep this as a simple key-value DTO for arbitrary parameters.

That would mean that the code below would instead of
```php
if (!$elements->hasReferenceLike($referenceName)) {
    continue;
}
$setReferencesCommands[] = SetNodeReferences::create(
    $commands->first->contentStreamId,
    $commands->first->nodeAggregateId,
    $commands->first->originDimensionSpacePoint,
      ReferenceName::fromString($referenceName),
      NodeReferencesToWrite::fromNodeAggregateIds($elements->getReferenceLike($referenceName))
);
```
be s.th. like:
```php
if (!$parameters->has($referenceName)) {
    continue;
}

$setReferencesCommands[] = SetNodeReferences::create(
    $commands->first->contentStreamId,
    $commands->first->nodeAggregateId,
    $commands->first->originDimensionSpacePoint,
      ReferenceName::fromString($referenceName),
      NodeReferencesToWrite::fromNodeAggregateIds(NodeAggregateIds::fromArrray($parameters->get($referenceName)),
);
```

(potentially with some error handling if the expected format does not match)

And I would even leave out the deserialization at this point..

Reply:
Yes thanks i had a little blockade in my brain as i somehow wanted to port the idea of separating properties and references https://github.com/neos/neos-development-collection/pull/4677 into the elements as well.

I also talked to Wilhelm about that and we found that `references` might be a little misleading as type here but introducing `nodeAggregateIds` or anything else is not worth it.
Its just important to note that the `type references` thingies have at that stage nothing in common with real reference edges. There is just a node editor that allows to "references" a node by its id.

I documented it like 

> The naming `references` in the `element` configuration does not refer to the content repository reference edges.
> Referring to a node will just denote that an editor will be used capable of returning node ids.
> The node ids might be used for setting references but that is up to a node-creation-handler.

So yes i simplified the vo and now it contains all elements in one blob.

> And I would even leave out the deserialisation at this point.

Here i have a strong objection. The data in a format only the neos ui understands. If every handler serialises it differently it leads to bugs like https://github.com/neos/neos-ui/issues/3719
By encapsulating it we make it an implementation detail and are later-on easier able to change the neos ui serialisation format.